modules/annex/annex_archive.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package annex
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"code.gitea.io/gitea/modules/git"
+)
+
+func repoArchiveTargz(repo *git.Repository, target io.Writer, prefix, commitID string) error {
+	// create a plain git tar archive
+	var tarB bytes.Buffer
+	err := repo.CreateArchive(repo.Ctx, git.TAR, &tarB, prefix != "", commitID)
+	if err != nil {
+		return err
+	}
+
+	gitFilesR := tar.NewReader(&tarB)
+
+	gzipW := gzip.NewWriter(target)
+	defer gzipW.Close()
+	tarW := tar.NewWriter(gzipW)
+	defer tarW.Close()
+
+	tree, err := repo.GetTree(commitID)
+	if err != nil {
+		return err
+	}
+
+	for {
+		oldHeader, err := gitFilesR.Next()
+		// TODO: handle non-local names in tar archives?
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+
+		// default to copying the current file from the archive created by repo.CreateArchive
+		header := oldHeader
+		dataR := io.Reader(gitFilesR)
+
+		// if we can get a annex content location for the file we use that instead
+		blob, err := tree.GetBlobByPath(strings.TrimPrefix(oldHeader.Name, prefix))
+		if err == nil {
+			annexPath, err := ContentLocation(blob)
+			if err == nil {
+				// blob corresponds to an annexed file
+
+				// build a tar header for the annexed file
+				file, err := os.Open(annexPath)
+				if err != nil {
+					return fmt.Errorf("opening %s failed: %w", annexPath, err)
+				}
+				stat, err := file.Stat()
+				if err != nil {
+					return fmt.Errorf("getting FileInfo for %s failed: %w", file.Name(), err)
+				}
+				header, err = tar.FileInfoHeader(stat, "")
+				if err != nil {
+					return fmt.Errorf("creating header failed: %w", err)
+				}
```

Comment:
:thinking: should this be the stat of the file in .annex/objects or the stat of the pointer file as known to git?

If the pointer file is a symlink, then it won't have a useful stat, and you have to rely on the file in .annex/objects, but if it's a git-smudge type file then it will. I'm just wondering because https://git-annex.branchable.com/internals/lockdown/ makes the permissions in .annex/objects/ different than you might expect, and, if users are paying attention, they should be expecting the perms as they seem them in their checkout.

I'll try to test for this manually when I deploy your branch.

Reply:
Currently it always uses the stat from .annex/objects which means annexed files e.g. get `-r--r--r--`. That might indeed be unexpected.